client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx

```diff
@@ -107,17 +112,23 @@ class EditContactInfoFormCard extends Component {
 	}
 
 	validate = ( fieldValues, onComplete ) => {
-		wpcom.validateDomainContactInformation(
-			fieldValues,
-			[ this.props.selectedDomain.name ],
-			( error, data ) => {
-				if ( error ) {
-					onComplete( error );
-				} else {
-					onComplete( null, data.messages || {} );
-				}
-			}
-		);
+		wp.req
+			.post(
+				'/me/domain-contact-information/validate',
+				mapRecordKeysRecursively(
+					{
+						contactInformation: fieldValues,
+						domainNames: [ this.props.selectedDomain.name ],
+					},
+					camelToSnakeCase
+				)
```

Comment:
I'm not very happy about `domains` importing an unrelated package like `@automattic/wpcom-checkout`. Could the functions be in `js-utils`?

Also, I suspect that a generic `mapRecordKeysRecursively` might not be needed. In the POST data, we can specify them as `{ contact_information, domain_names }` directly, can't we?

And what remains is always the same data structure, `contact_information`. That one is flat, with no nested fields, and we could have a custom `fromApi`/`toApi` function that enumerates the fields explicitely. We already have many conversion helpers like this.

Reply:
> I'm not very happy about domains importing an unrelated package like @automattic/wpcom-checkout. Could the functions be in js-utils?

This has been brought up before for other methods, so I think it makes sense. I could spin up a PR for that.

> Also, I suspect that a generic mapRecordKeysRecursively might not be needed. In the POST data, we can specify them as { contact_information, domain_names } directly, can't we?

Yes, however, `contact_information` contains subfields and their keys need to be recursively mapped when retrieving and before sending to the relevant endpoints.

Reply:
> I'm not very happy about domains importing an unrelated package like @automattic/wpcom-checkout. Could the functions be in js-utils?

Given that we need `mapRecordKeysRecursively`, I wonder if you could say the same for the `mapRecordKeysRecursively` method - would you expect to see it in `@automatic/js-utils` as well, @jsnajdr?

Reply:
> I wonder if you could say the same for the `mapRecordKeysRecursively` method

At this moment, I'm still convinced we don't need it 🙂 

> however, `contact_information` contains subfields and their keys need to be recursively mapped

Do they really? I tried editing contact info for one of my domains and the `contact_information` structure is flat. Set of fields where the values are always strings.

Reply:
> Do they really? I tried editing contact info for one of my domains and the contact_information structure is flat.

It's not always flat. Try with a `.ca` or `.fr` or `.uk` domain. You'll see nested fields for those:

![](https://cldup.com/sEye3LqMyp.png)

Reply:
I see, there's the `extra` field 🙍‍♂️ Still, the conversion is done only for the "contact information" structure. Then it would be more acceptable, even though not ideal, if the `@automattic/wpcom-checkout` exported functions specific for that:
- `contactInformationFromREST()`
- `contactInformationToREST()`
Implemented internally as some recursive `Object.fromEntries` stuff.

Reply:
Considering that the PR is not introducing `mapRecordKeysRecursively` and it pre-exists, are these contact-specific functions something you insist to see in this PR, or is it something we can get to incrementally? After all, the purpose of this PR is to reuse as much as possible from what we have already in order to migrate away from `wpcom.undocumented()`, not to introduce new APIs or improve existing ones. 

I'd also love to hear @sirbrillig's thoughts about this.

Reply:
It's worth noting that `mapRecordKeysRecursively` and those `camelCase` / `snakeCase` functions are all pretty generic and might make sense to belong to a more generic package like `@automattic/js-utils`. Won't that resolve your concerns, @jsnajdr?

Reply:
> Won't that resolve your concerns?

Yes, there's a family of Lodash-inspired object-mapping functions, `mapKeys`, `mapValues`, and `mapKeysRecursively`, that we use a lot and will need to migrate one day. `js-utils` is a good place for them.

`camelCase` and `snakeCase` are another set of helpers that will need to be migrated away from Lodash.

Reply:
> The only thing that I think should be changed in this PR are the POST data in the validate requests:

Oh, I think I've been misunderstanding you. That makes sense, will update it in the next iteration.

>  Which structures are we really mapping with these functions?

It's worth noting that this PR isn't set to answer that kind of questions at all. But we're using generic functions and it's pretty straightforward to trace what we're mapping and how it ends up mapped IMHO.

> Yes, there's a family of Lodash-inspired object-mapping functions

Is this something we can follow-up with? I'm happy to move all those after we merge this PR. I don't think the move necessarily has to block this PR. What do you think?

Reply:
> What do you think?

I think the current situation with the mapping functions is not ideal, but we don't need to fix it all in one PR. This `wpcom.undocumented` migration is ready to go, the rest can be independent follow-ups.

Reply:
We do know the [expected format of the contact details](https://github.com/Automattic/wp-calypso/blob/c3fb2139b727879145fa4f747718bb71bed43fd5/packages/wpcom-checkout/src/types.ts#L388) and we have [a function that can create it explicitly](https://github.com/Automattic/wp-calypso/blob/c3fb2139b727879145fa4f747718bb71bed43fd5/client/my-sites/checkout/composite-checkout/types/wpcom-store-state.ts#L505) from the camelCase we use in checkout, but I didn't use it when migrating away from `undocumented` for the transactions endpoint because we haven't yet typed the other data sent to the transactions endpoint. Eventually I'd like to get rid of the case changes and mapping in checkout entirely. In theory, those functions could work well here, though.

The biggest problem is probably that the domain contact forms here still use `formStateController` and its input/output is probably not compatible with the conversion functions I mentioned. We migrated away from that class in checkout but as these forms are not checkout related we never touched them. Still, there is a precedent for that work and it could be done as a separate project.

Reply:
Thank you both!

Reply:
Moved the methods to `js-utils` in https://github.com/Automattic/wp-calypso/pull/58110